### PR TITLE
[Downloads] File list

### LINKED ISSUE #35142: [Downloads] File list
**Job story**

- When I put my digital product or service on sale on my website,
- I want to manage basic file settings
- So that the customer has a better post-purchase experience

**Design**

AkVNGImLgSqCObTQ3idVn7-fi-7001_308059

**Acceptance criteria**

- [x] When the user uploads at least 1 file, the product gets the Downloadable trait on the backend (the equivalent of checking the Downloadable box in the current UX).
- [x] By default, the list displays the files by their actual names (including the extension/format).
- [x] If the file has a custom name added by the user, we show it first. The actual file name is then shown in a second line.
- [x] Each file is a separate item in the list with action X (remove).
- [x] The user can click the X button to remove a file from the list (but not delete it from the library).

### PR DESCRIPTION
**Pull request by mdperez86**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

Closes #35142

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Make sure feature `New product editor` is enabled under `/wp-admin/admin.php?page=wc-settings&tab=advanced&section=features`
2. Make sure feature `product-virtual-downloadable` is enabled under `Features` tab from `/wp-admin/tools.php?page=woocommerce-admin-test-helper` (WooCommerce Beta Tester plugin) is required
3. Go to `/wp-admin/admin.php?page=wc-admin&path=/add-product`
4. Under `General` tab and the bottom of the page a new `Downloads` section should be shown.
5. By default, the list displays the files by their actual names (including the extension/format) after uploading one or many files.
6. Each file is a separate item in the list with action X (remove). When click it the file should be removed from the product but not from the media library. 

https://github.com/woocommerce/woocommerce/assets/13334210/78490b71-6228-4b9d-8932-83f9054537c0

### Changelog entry

-   [ ] Automatically create a changelog entry from the details below.

<details>

#### Significance

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Comment 

</details>

### CODE REVIEW

File: packages/js/product-editor/src/blocks/downloads/edit.tsx [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+							</ListItem>
+						);
+					} ) }
+				</Sortable>
+			) : (
+				<MediaUploader
+					label={
+						<>
+							<UploadImage />
+							<p className="woocommerce-product-form__remove-files-drop-zone-label">
```
Comment (octaedro) [inline packages/js/product-editor/src/blocks/downloads/edit.tsx]: Can we rename this className to `woocommerce-product-form__supported-files-drop-zone-label`?

File: packages/js/product-editor/src/blocks/downloads/editor.scss [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+		justify-content: center;
+
+		.woocommerce-media-uploader__label {
+			display: flex;
+			align-items: center;
+			justify-content: center;
+			gap: $grid-unit-40;
+			margin-bottom: 0;
+
+			.woocommerce-product-form__remove-files-drop-zone-label {
```
Comment (octaedro) [inline packages/js/product-editor/src/blocks/downloads/editor.scss]: Same here

File: plugins/woocommerce/src/Admin/Features/ProductBlockEditor/ProductTemplates/SimpleProductTemplate.php [hunk 3]
```diff
@@ -248,6 +248,26 @@ private function add_general_group_blocks() {
...
+					'description' => __( "Add any files you'd like to make available for the customer to download after purchasing, such as instructions or warranty info.", 'woocommerce' ),
+				],
+			]
+		)->add_block(
+			[
+				'id'         => 'product-downloads',
+				'blockName'  => 'woocommerce/product-downloads-field',
+				'order'      => 10,
+				'attributes' => [
+					'images' => [],
```
Comment (octaedro) [inline plugins/woocommerce/src/Admin/Features/ProductBlockEditor/ProductTemplates/SimpleProductTemplate.php]: Should this be `downloads`?

File: packages/js/product-editor/src/blocks/downloads/block.json [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+	"$schema": "https://schemas.wp.org/trunk/block.json",
+	"apiVersion": 2,
+	"name": "woocommerce/product-downloads-field",
+	"title": "Product downloads",
+	"category": "widgets",
+	"description": "The product downloads.",
+	"keywords": [ "products", "downloads" ],
+	"textdomain": "default",
+	"attributes": {
+		"name": {
```
Comment (octaedro) [inline packages/js/product-editor/src/blocks/downloads/block.json]: I think that we're not using the [attribute 'name'](https://github.com/woocommerce/woocommerce/pull/40526/files#diff-4d3fbd25af47c29cf34a06bd1ba13f25dba78e1a3f19034c1a67c28c89a3dc58R267), right?

**Review by octaedro — COMMENTED** [review]
You did a fantastic job with the Downloads file list @mdperez86! Overall, this is testing well. However, after creating a downloadable product, if I reload the page, I'm redirected to the old editor. I've also left a couple of small comments.

**Comment by octaedro** [comment]
> after creating a downloadable product, if I reload the page, I'm redirected to the old editor

I think that [this line](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/src/Admin/Features/ProductBlockEditor/RedirectionController.php#L65) is the culprit of the redirection after reloading the page.

File: packages/js/product-editor/src/blocks/downloads/block.json (hunk 4 above)
Comment (mdperez86) [reply packages/js/product-editor/src/blocks/downloads/block.json]: Nope but it's required to set `"__experimentalRole": "content"` which will allow to select text under the block.

File: packages/js/product-editor/src/blocks/downloads/edit.tsx (hunk 1 above)
Comment (mdperez86) [reply packages/js/product-editor/src/blocks/downloads/edit.tsx]: done -> https://github.com/woocommerce/woocommerce/pull/40526/commits/82381641356554dabd2c8794593f35ec28857997

File: packages/js/product-editor/src/blocks/downloads/editor.scss (hunk 2 above)
Comment (mdperez86) [reply packages/js/product-editor/src/blocks/downloads/editor.scss]: done -> https://github.com/woocommerce/woocommerce/pull/40526/commits/82381641356554dabd2c8794593f35ec28857997

File: plugins/woocommerce/src/Admin/Features/ProductBlockEditor/ProductTemplates/SimpleProductTemplate.php (hunk 3 above)
Comment (mdperez86) [reply plugins/woocommerce/src/Admin/Features/ProductBlockEditor/ProductTemplates/SimpleProductTemplate.php]: done -> https://github.com/woocommerce/woocommerce/pull/40526/commits/82381641356554dabd2c8794593f35ec28857997

**Comment by mdperez86** [comment]
> > after creating a downloadable product, if I reload the page, I'm redirected to the old editor
> 
> I think that [this line](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/src/Admin/Features/ProductBlockEditor/RedirectionController.php#L65) is the culprit of the redirection after reloading the page.

Talking with @AnnaMag we should hide the whole functionality behind a feature flag, so we are going to need an issue for that.

**Review by octaedro — APPROVED** [review]
Thank you @mdperez86 for addressing the changes. LGTM 🚀

**Pull request merged**